File: 0688-knight-probability-in-chessboard/0688-knight-probability-in-chessboard.cpp
class Solution {
public:
    double dp[26][26][102];
    
    int xMove[8] = { -2, -2, -1, -1, 1, 1, 2, 2 };
    int yMove[8] = { -1, 1, -2, 2, -2, 2, -1, 1 };
    
    double solve(int row, int col, int n, int k) {
        if (row < 0 || col < 0 || row >= n || col >= n) {
            return 0;
        }
        if (k == 0) {
            return 1;
        }
        if (dp[row][col][k] != -1.0) {
            return dp[row][col][k];
        }
        double res = 0;
        for (int i = 0; i < 8; i++) {
            res += solve(row + xMove[i], col + yMove[i], n, k - 1);
        }
        return dp[row][col][k] = res;
    }

    double knightProbability(int n, int k, int row, int column) {
        for (int i = 0; i < 26; i++) {
            for (int j = 0; j < 26; j++) {
                for (int k = 0; k < 102; k++) {
                    dp[i][j][k] = -1.0;
                }
            }
        }
        double favourableOutcomes = solve(row, column, n, k);
        double totalOutcomes = pow(8.0, k);
        return favourableOutcomes / totalOutcomes;
    }
};

File: 0688-knight-probability-in-chessboard/0688-knight-probability-in-chessboard.java
class Solution {
    public double knightProbability(int n, int k, int row, int col) {
        Double[][][] dp = new Double[n][n][k + 1];
        return solve(row, col, k, n, dp);
    }
    
    public double solve(int row, int col, int k, int n, Double[][][] dp) {
        if (row < 0 || col < 0 || row >= n || col >= n || k < 0) {
            return 0.0;
        }
        if (k == 0) {
            return 1.0;
        }
        if (dp[row][col][k] != null) {
            return dp[row][col][k];
        }
        return dp[row][col][k] =
            solve(row - 2, col - 1, k - 1, n, dp) / 8.0 +
            solve(row - 1, col - 2, k - 1, n, dp) / 8.0 +
            solve(row + 1, col - 2, k - 1, n, dp) / 8.0 +
            solve(row + 2, col - 1, k - 1, n, dp) / 8.0 +
            solve(row + 2, col + 1, k - 1, n, dp) / 8.0 +
            solve(row + 1, col + 2, k - 1, n, dp) / 8.0 +
            solve(row - 1, col + 2, k - 1, n, dp) / 8.0 +
            solve(row - 2, col + 1, k - 1, n, dp) / 8.0;
    }
}